libs/debugging/include/hpx/debugging/print.hpp

```diff
@@ -225,71 +307,201 @@ namespace hpx { namespace debug {
             tuple_print(os, tup, std::make_index_sequence<sizeof...(Args)>());
         }
 
+#else
+        // C++14 version
+        // helper function to print a tuple of any size
+        template <typename TupleType, std::size_t N>
+        struct tuple_printer
+        {
+            static void print(std::ostream& os, const TupleType& t)
+            {
+                tuple_printer<TupleType, N - 1>::print(os, t);
+                os << " " << std::get<N - 1>(t);
+            }
+        };
+
+        template <typename TupleType>
+        struct tuple_printer<TupleType, 1>
+        {
+            static void print(std::ostream& os, const TupleType& t)
+            {
+                os << std::get<0>(t);
+            }
+        };
+
+        template <class... Args>
+        void tuple_print(std::ostream& os, const std::tuple<Args...>& t)
+        {
+            tuple_printer<decltype(t), sizeof...(Args)>::print(os, t);
+        }
+
+        template <typename Arg, typename... Args>
+        void variadic_print(
+            std::ostream& os, const Arg& arg, const Args&... args)
+        {
+            os << arg;
+            using expander = int[];
+            (void) expander{0, (void(os << ' ' << args), 0)...};
+        }
+
+        template <class... Args>
+        void tuple_print(std::ostream& os, const Args&... args)
+        {
+            variadic_print(os, args...);
+        }
+
+#endif
+
+    }    // namespace detail
+
+    namespace detail {
+
+        // ------------------------------------------------------------------
+        // helper class for printing time since start
+        // ------------------------------------------------------------------
+        struct hostname_print_helper
+        {
+            const char *get_hostname() const {
+                static bool initialized = false;
+                static char hostname_[20];
+                if (!initialized) {
+                    initialized = true;
+                    gethostname(hostname_, std::size_t(12));
+                    std::string temp = "(" + std::to_string(guess_rank()) + ")";
+                    std::strcat(hostname_, temp.c_str());
+                }
+                return hostname_;
+            }
+
+            int guess_rank() const {
+#ifdef DEBUGGING_PRINT_LINUX
```

Comment:
WHat's the rationale of limiting this to `LINUX`?

Reply:
I didn't want to add and env var checking things that might not compile on windows without testing them. I'm guessing that I'm the only person using this debug print stuff, so I didn't think anyone would notice.

Reply:
The code you wrapped should be working fine on other platforms as long as you use:
```
#ifdef __APPLE__
#include <crt_externs.h>
#define environ (*_NSGetEnviron())
#elif defined(HPX_WINDOWS)
#define environ _environ
#else
extern char** environ;
#endif
```